# Подход 1: градиентный бустинг "в лоб"
Один из самых универсальных алгоритмов, изученных в нашем курсе, является градиентный бустинг. Он не очень требователен к данным, восстанавливает нелинейные зависимости, и хорошо работает на многих наборах данных, что и обуславливает его популярность. Вполне разумной мыслью будет попробовать именно его в первую очередь.

1. Считайте таблицу с признаками из файла features.csv с помощью кода, приведенного выше.

### Загружаем и разархивируем features.zip — архив с обучающей выборкой

In [ ]:
try:
    import wget
except:
    !pip install wget > nul
    import wget
import zipfile
zip_file = wget.download('https://d3c33hcgiwev3.cloudfront.net/_47ea8666d262362a0b41164cb497658e_features.zip?Expires=1608595200&Signature=fMbNpjE8hxxRyYp6V7E8cQlsDVzvN~~JLILkv8rEmDUL9xEQfUAJ-wa8AA560QrosdOtkkvJ6A-QE6Tn1pM5qTfv2kFTQpYblx2hoqH6VsO5vGZdhhhiUEoUO7Pe0Rewy8OvSzGEkSwL6b2av3Ar8bVxK~h-wXY7lE-1f2OK22A_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A','./features.zip')
with zipfile.ZipFile(zip_file,"r") as f:
  f.extractall()

In [ ]:
import pandas as pd
data_train = pd.read_csv('./features.csv', index_col='match_id')
data_train

start_time  lobby_type  ...  barracks_status_radiant  barracks_status_dire
match_id                          ...                                               
0         1430198770           7  ...                       51                     0
1         1430220345           0  ...                       63                     1
2         1430227081           7  ...                        0                    63
3         1430263531           1  ...                       50                    63
4         1430282290           7  ...                        3                    63
...              ...         ...  ...                      ...                   ...
114402    1450265551           1  ...                       48                    63
114403    1450277704           0  ...                       63                     3
114404    1450291848           1  ...                       51                    63
114405    1450292986           1  ...                       63                    63
114406    1450313370           7  ...                       63                     0

[97230 rows x 108 columns]

### Загружаем и разархивируем features_test.zip — архив с тестовой выборкой

In [ ]:
zip_file_test = wget.download('https://d3c33hcgiwev3.cloudfront.net/_6866d458468373e164b1e859f3103826_features_test.zip?Expires=1608595200&Signature=OHYPh5u17GdTOxLav9LBwAfh66CMcAuzMd0u04vyGrD0hrMsvS2DrSNJEku5VJSC3PT5DaLKCXR1JCqijvOGleftRw8d6JonTab6K-SHIlqOGYDnYoGw5kgH7FxfdVw05F3Quo8n~XYreOkSyvijueAxsHl8rqEXCxwTQjz1o5c_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A','./features_test.zip')
with zipfile.ZipFile(zip_file_test,"r") as f_test:
  f_test.extractall()
data_test = pd.read_csv('./features_test.csv', index_col='match_id')
data_test

start_time  lobby_type  ...  dire_ward_sentry_count  dire_first_ward_time
match_id                          ...                                              
6         1430287923           0  ...                       0                 118.0
7         1430293357           1  ...                       1                  16.0
10        1430301774           1  ...                       0                 -34.0
13        1430323933           1  ...                       1                 -26.0
16        1430331112           1  ...                       1                 -33.0
...              ...         ...  ...                     ...                   ...
114369    1450212780           7  ...                       1                 -33.0
114377    1450222875           1  ...                       1                 -18.0
114378    1450223593           1  ...                       0                 -36.0
114393    1450244771           0  ...                       0                   NaN
114398    1450255429           1  ...                       0                  27.0

[17177 rows x 102 columns]

In [ ]:
train_Y=data_train['radiant_win']

Удалите признаки, связанные с итогами матча (они помечены в описании данных как отсутствующие в тестовой выборке).

In [ ]:
columns_train_difference=data_train.columns.difference(data_test.columns.values.tolist()).tolist()
data_train.drop(columns_train_difference, axis=1, inplace=True)
data_train

start_time  lobby_type  ...  dire_ward_sentry_count  dire_first_ward_time
match_id                          ...                                              
0         1430198770           7  ...                       2                 -52.0
1         1430220345           0  ...                       1                  -5.0
2         1430227081           7  ...                       1                  13.0
3         1430263531           1  ...                       0                  27.0
4         1430282290           7  ...                       0                 -16.0
...              ...         ...  ...                     ...                   ...
114402    1450265551           1  ...                       0                 -24.0
114403    1450277704           0  ...                       0                 -17.0
114404    1450291848           1  ...                       1                 -15.0
114405    1450292986           1  ...                       1                 -42.0
114406    1450313370           7  ...                       0                 -39.0

[97230 rows x 102 columns]

2. Проверьте выборку на наличие пропусков с помощью функции count(), которая для каждого столбца показывает число заполненных значений. Много ли пропусков в данных? Запишите названия признаков, имеющих пропуски, и попробуйте для любых двух из них дать обоснование, почему их значения могут быть пропущены.

In [ ]:
train_size=len(data_train)
print(f"Количество строк: {train_size}" )
for col in data_train.columns.values.tolist():
    count=data_train[col].count()
    if count!=train_size:
        print(f"Признак: {col}, значений: {count}")

Количество строк: 97230
Признак: first_blood_time, значений: 77677
Признак: first_blood_team, значений: 77677
Признак: first_blood_player1, значений: 77677
Признак: first_blood_player2, значений: 53243
Признак: radiant_bottle_time, значений: 81539
Признак: radiant_courier_time, значений: 96538
Признак: radiant_flying_courier_time, значений: 69751
Признак: radiant_first_ward_time, значений: 95394
Признак: dire_bottle_time, значений: 81087
Признак: dire_courier_time, значений: 96554
Признак: dire_flying_courier_time, значений: 71132
Признак: dire_first_ward_time, значений: 95404


3. Замените пропуски на нули с помощью функции fillna(). На самом деле этот способ является предпочтительным для логистической регрессии, поскольку он позволит пропущенному значению не вносить никакого вклада в предсказание. Для деревьев часто лучшим вариантом оказывается замена пропуска на очень большое или очень маленькое значение — в этом случае при построении разбиения вершины можно будет отправить объекты с пропусками в отдельную ветвь дерева. Также есть и другие подходы — например, замена пропуска на среднее значение признака. Мы не требуем этого в задании, но при желании попробуйте разные подходы к обработке пропусков и сравните их между собой.

In [ ]:
data_train.fillna(0, method=None, axis=1, inplace=True)

4. Какой столбец содержит целевую переменную? Запишите его название.

### Целевая переменная 'radiant_win'- 1, если победила команда Radiant, 0 — иначе

5. Забудем, что в выборке есть категориальные признаки, и попробуем обучить градиентный бустинг над деревьями на имеющейся матрице "объекты-признаки". Зафиксируйте генератор разбиений для кросс-валидации по 5 блокам (KFold), не забудьте перемешать при этом выборку (shuffle=True), поскольку данные в таблице отсортированы по времени, и без перемешивания можно столкнуться с нежелательными эффектами при оценивании качества. Оцените качество градиентного бустинга (GradientBoostingClassifier) с помощью данной кросс-валидации, попробуйте при этом разное количество деревьев (как минимум протестируйте следующие значения для количества деревьев: 10, 20, 30). Долго ли настраивались классификаторы? Достигнут ли оптимум на испытанных значениях параметра n_estimators, или же качество, скорее всего, продолжит расти при дальнейшем его увеличении?

In [ ]:
import time
import datetime
from sklearn.model_selection import KFold
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import cross_val_score

kf = KFold(n_splits=5, random_state=42, shuffle=True)
n_estimators = {10, 20, 30, 100}

In [ ]:
for est in n_estimators:
  start_time = datetime.datetime.now()
  clf = GradientBoostingClassifier(n_estimators=est)
  scores = cross_val_score(clf, data_train, train_Y, scoring='roc_auc', cv=kf)
  print ('Time elapsed:', datetime.datetime.now() - start_time)
  mean_score = round(scores.mean()*100,2)
  print(f"Количество деревьев {est}, качество roc_auc на кросс-валидации по тренировочной выборке: {mean_score}%")

Time elapsed: 0:09:04.645642
Количество деревьев 100, качество roc_auc на кросс-валидации по тренировочной выборке: 70.62%
Time elapsed: 0:00:55.755712
Количество деревьев 10, качество roc_auc на кросс-валидации по тренировочной выборке: 66.49%
Time elapsed: 0:01:51.532591
Количество деревьев 20, качество roc_auc на кросс-валидации по тренировочной выборке: 68.25%
Time elapsed: 0:02:47.356322
Количество деревьев 30, качество roc_auc на кросс-валидации по тренировочной выборке: 69.0%


#ОТЧЕТ по подходу №1
## 1.Пропуски имеют следующие признаки:
* first_blood_time: игровое время первой крови
* first_blood_team: команда, совершившая первую кровь (0 — Radiant, 1 — Dire)
* first_blood_player1: игрок, причастный к событию
* first_blood_player2: второй игрок, причастный к событию
* radiant_bottle_time: время первого приобретения командой предмета "bottle"
* radiant_courier_time: время приобретения предмета "courier"
* radiant_flying_courier_time: время приобретения предмета "flying_courier
* radiant_first_ward_time: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля
* dire_bottle_time: время первого приобретения командой предмета "bottle"
* dire_courier_time: время приобретения предмета "courier"
* dire_flying_courier_time: время приобретения предмета "flying_courier
* dire_first_ward_time: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля
### Отсутствующие данные, скорее всего, свидетельствуют от том, что данные события не призошли в течение 5 минут игры.
## 2.'radiant_win'
## 3. C 30 деревьями кросс-валидация выполнялась ~2.30 мин.
## 4. В данном случае увеличивать деревья больше 30  имеет смысл при наличии достаточного времени или вычислительных мощностей. Максимальное количество деревьев можно подобрать с помощью GridSearchCV, но это довольно ресурсоемкий метод. При 100 деревьях качество увеличивается менее чем на 2%, а время подсчета почти в 4 раза. Для ускорения процесса при увеличении числа деревьев следует уменьшать глубину деревьев и/или использовать только часть признаков.


### Подход 2: логистическая регрессия

Линейные методы работают гораздо быстрее композиций деревьев, поэтому кажется разумным воспользоваться именно ими для ускорения анализа данных. Одним из наиболее распространенных методов для классификации является логистическая регрессия.

**Важно:** не забывайте, что линейные алгоритмы чувствительны к масштабу признаков! Может пригодиться sklearn.preprocessing.StandartScaler.

1. Оцените качество логистической регрессии (sklearn.linear_model.LogisticRegression с L2-регуляризацией) с помощью кросс-валидации по той же схеме, которая использовалась для градиентного бустинга. Подберите при этом лучший параметр регуляризации (C). Какое наилучшее качество у вас получилось? Как оно соотносится с качеством градиентного бустинга? Чем вы можете объяснить эту разницу? Быстрее ли работает логистическая регрессия по сравнению с градиентным бустингом?

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
import pandas as pd

Воспользуемся StandardScaler

In [ ]:
data_train_sc=pd.DataFrame(data=StandardScaler().fit_transform(data_train))
data_train_sc

0         1         2    ...       99        100       101
0     -2.544364  1.540688 -1.244228  ... -0.551154  1.846004 -1.121494
1     -2.540452 -0.927798 -0.292258  ...  0.678170  0.437788  0.043947
2     -2.539231  1.540688 -0.568637  ...  0.678170  0.437788  0.490286
3     -2.532622 -0.575157 -0.691471  ... -0.551154 -0.970428  0.837439
4     -2.529221  1.540688 -1.182811  ...  0.678170 -0.970428 -0.228816
...         ...       ...       ...  ...       ...       ...       ...
97225  1.093978 -0.575157 -0.138714  ...  0.678170 -0.970428 -0.427189
97226  1.096181 -0.927798 -0.261549  ... -0.551154 -0.970428 -0.253612
97227  1.098746 -0.575157  1.427430  ...  0.678170  0.437788 -0.204019
97228  1.098952 -0.575157  1.488848  ...  0.678170  0.437788 -0.873528
97229  1.102648  1.540688 -0.046588  ... -0.551154 -0.970428 -0.799138

[97230 rows x 102 columns]

In [ ]:
%%time
lr = LogisticRegression(n_jobs = -1);
grid = {'C': [0.001, 0.01, 0.05 ,0.1, 1, 5]} 
cv = KFold(n_splits=5, random_state=42, shuffle=True) #Кроссвалидация
gs = GridSearchCV(lr, grid, scoring='roc_auc', cv=cv, verbose=0)
gs.fit(data_train_sc, train_Y)
print('Лучший результат:', round(gs.best_score_*100,2))
print('Лучшие параметры:', gs.best_params_)

Лучший результат: 71.65
Лучшие параметры: {'C': 0.01}
CPU times: user 3.79 s, sys: 4.1 s, total: 7.89 s
Wall time: 2min 58s



2. Среди признаков в выборке есть категориальные, которые мы использовали как числовые, что вряд ли является хорошей идеей. Категориальных признаков в этой задаче одиннадцать: lobby_type и r1_hero, r2_hero, ..., r5_hero, d1_hero, d2_hero, ..., d5_hero. Уберите их из выборки, и проведите кросс-валидацию для логистической регрессии на новой выборке с подбором лучшего параметра регуляризации. Изменилось ли качество? Чем вы можете это объяснить?

In [ ]:
cat_features = ['r%s_hero' % i for i in range(1, 6)]+['d%s_hero' % i for i in range(1, 6)]#колонки героев
cat_features.append('lobby_type')
data_train_new = data_train.drop(cat_features, axis=1)
data_train_new_sc=pd.DataFrame(data=StandardScaler().fit_transform(data_train_new))
data_train_new_sc

0         1         2   ...        88        89        90
0     -2.544364  1.400808  1.525972  ... -0.551154  1.846004 -1.121494
1     -2.540452  0.501314 -0.080139  ...  0.678170  0.437788  0.043947
2     -2.539231  0.501314  0.151070  ...  0.678170  0.437788  0.490286
3     -2.532622  0.501314  0.962950  ... -0.551154 -0.970428  0.837439
4     -2.529221  0.501314  0.348745  ...  0.678170 -0.970428 -0.228816
...         ...       ...       ...  ...       ...       ...       ...
97225  1.093978  0.501314  0.834109  ...  0.678170 -0.970428 -0.427189
97226  1.096181  0.501314  0.987660  ... -0.551154 -0.970428 -0.253612
97227  1.098746  0.501314  0.292267  ...  0.678170  0.437788 -0.204019
97228  1.098952 -0.398181 -0.173682  ...  0.678170  0.437788 -0.873528
97229  1.102648 -0.398181 -0.318409  ... -0.551154 -0.970428 -0.799138

[97230 rows x 91 columns]

In [ ]:
%%time
gs = GridSearchCV(lr, grid, scoring='roc_auc', cv=cv, verbose=0)
gs.fit(data_train_new_sc, train_Y)
print('Лучший результат:', round(gs.best_score_*100,2))
print('Лучшие параметры:', gs.best_params_)

Лучший результат: 71.66
Лучшие параметры: {'C': 0.01}
CPU times: user 3.55 s, sys: 3.89 s, total: 7.44 s
Wall time: 2min 30s


3. На предыдущем шаге мы исключили из выборки признаки rM_hero и dM_hero, которые показывают, какие именно герои играли за каждую команду. Это важные признаки — герои имеют разные характеристики, и некоторые из них выигрывают чаще, чем другие. Выясните из данных, сколько различных идентификаторов героев существует в данной игре (вам может пригодиться фукнция unique или value_counts).

In [ ]:
cat_features.remove('lobby_type')
N_hero=pd.Series(data_train[cat_features].values.flatten()).drop_duplicates().shape[0]
print(f'В игре: {N_hero} героев')

В игре: 108 героев


4. Воспользуемся подходом "мешок слов" для кодирования информации о героях. Пусть всего в игре имеет N различных героев. Сформируем N признаков, при этом i-й будет равен нулю, если i-й герой не участвовал в матче; единице, если i-й герой играл за команду Radiant; минус единице, если i-й герой играл за команду Dire. Ниже вы можете найти код, который выполняет данной преобразование. Добавьте полученные признаки к числовым, которые вы использовали во втором пункте данного этапа.

In [ ]:
#Функция ix устарела, воспользуемся loc

import numpy as np

X_pick = np.zeros((data_train.shape[0], N_hero))
 
for i, match_id in enumerate(data_train.index):
    for p in range(5):
        X_pick[i, data_train.loc[match_id, 'r%d_hero' % (p+1)]-N_hero] = 1
        X_pick[i, data_train.loc[match_id, 'd%d_hero' % (p+1)]-N_hero] = -1

In [ ]:
#Добавляем "Мешoк слов" к отмасштабированной выбрке числовых признаков, без категориальных 
full_data = np.hstack([np.array(data_train_new_sc), X_pick])
full_data.shape

(97230, 199)

5. Проведите кросс-валидацию для логистической регрессии на новой выборке с подбором лучшего параметра регуляризации. Какое получилось качество? Улучшилось ли оно? Чем вы можете это объяснить?

In [ ]:
%%time
lr = LogisticRegression(n_jobs = -1);
grid = {'C': np.linspace(0.03, 0.1, num=8)} 
cv = KFold(n_splits=5, random_state=42, shuffle=True) #Кроссвалидация
gs = GridSearchCV(lr, grid, scoring='roc_auc', cv=cv, verbose=0)
gs.fit(full_data, train_Y)
print('Лучший результат:', round(gs.best_score_*100,2))
print('Лучшие параметры:', gs.best_params_)

Лучший результат: 74.99
Лучшие параметры: {'C': 0.05}
CPU times: user 5.83 s, sys: 6.38 s, total: 12.2 s
Wall time: 18min 2s


6. Постройте предсказания вероятностей победы команды Radiant для тестовой выборки с помощью лучшей из изученных моделей (лучшей с точки зрения AUC-ROC на кросс-валидации). Убедитесь, что предсказанные вероятности адекватные — находятся на отрезке [0, 1], не совпадают между собой (т.е. что модель не получилась константной).

In [ ]:
#Заменяем пропуски на "0"
data_test.fillna(0, method=None, axis=1, inplace=True)
#Выкидываем из тестовой выборки категориальные признаки
cat_features = ['r%s_hero' % i for i in range(1, 6)]+['d%s_hero' % i for i in range(1, 6)]#колонки героев
cat_features.append('lobby_type')
data_test_new = data_test.drop(cat_features, axis=1)
# Масштабируем признаки
data_test_new_sc=pd.DataFrame(data=StandardScaler().fit_transform(data_test_new))
data_test_new_sc

0         1         2   ...        88        89        90
0     -2.514875  0.474746 -0.237743  ... -0.571552 -0.983356  3.200079
1     -2.513895 -1.365630 -1.207913  ... -0.571552  0.433564  0.600867
2     -2.512377 -1.365630 -0.862057  ...  0.668978 -0.983356 -0.673257
3     -2.508381 -0.445442 -0.938323  ... -0.571552  0.433564 -0.469397
4     -2.507087  0.474746  0.038942  ... -0.571552  0.433564 -0.647774
...         ...       ...       ...  ...       ...       ...       ...
17172  1.078106  1.394934  1.448970  ... -0.571552  0.433564 -0.647774
17173  1.079926 -0.445442 -0.867378  ...  1.909508  0.433564 -0.265537
17174  1.080056 -1.365630 -1.174215  ...  1.909508 -0.983356 -0.724222
17175  1.083875  0.474746  1.076510  ... -1.812082 -0.983356  0.193147
17176  1.085797  0.474746 -0.039098  ... -0.571552 -0.983356  0.881174

[17177 rows x 91 columns]

In [ ]:
# Составляем "Мешок слов"
X_pick_test = np.zeros((data_test.shape[0], N_hero))
 
for i, match_id in enumerate(data_test.index):
    for p in range(5):
        X_pick[i, data_test.loc[match_id, 'r%d_hero' % (p+1)]-N_hero] = 1
        X_pick[i, data_test.loc[match_id, 'd%d_hero' % (p+1)]-N_hero] = -1
#Добавляем "Мешoк слов" к отмасштабированной выбрке числовых признаков ТЕСТОВОГО датасета
full_data_test = np.hstack([np.array(data_test_new_sc), X_pick_test])
full_data_test.shape

(17177, 199)

In [ ]:
#Обучаем модель с наилучшими параметрами по тренировочной выборке
start_time = datetime.datetime.now()
final_model = LogisticRegression(C = 0.05, n_jobs = -1).fit(full_data, train_Y)
#Предсказывает вероятность на тестовой выборке
y_pred = final_model.predict_proba(full_data_test)
print ('Time elapsed:', datetime.datetime.now() - start_time)

Time elapsed: 0:00:34.747701


In [ ]:
#Проверка неконстантности модели
print(y_pred)
print("\nMинимальное значение прогноза: ", y_pred[:, 1].min())
print("\nМаксимальное значение прогноза: ", y_pred[:, 1].max() )

[[0.48498959 0.51501041]
 [0.33889545 0.66110455]
 [0.6797318  0.3202682 ]
 ...
 [0.77309387 0.22690613]
 [0.63890743 0.36109257]
 [0.46961339 0.53038661]]

Mинимальное значение прогноза:  0.005744518567402761

Максимальное значение прогноза:  0.9937054595749831


#**Отчет по подходу 2: логистическая регрессия**
##1. Качество на кроссвалидации (без дополнительной обработки признаков) у логистической регрессии над всеми исходными признаками - 71.65%, что выше чем 70.62% у градиентного бустинга. Разницу можно объяснить особенностями  исходных данных.
## Скорость деревьев заметно падает с увеличением их числа. На финальном тесте видно, что логистическая регрессия работает значительно быстрее.
##2. При удалении категориальных признаков качество модели возросло на 0,01% и составило 71,66%. Линейные модели не предназначены для работы с категориальными признаками, эти признаки сорее всего просто вносят дополнительный шум.
##3. По моим расчетам, в игре 108 героев.
##4. При добавлении "мешка слов" по героям, качество улучшается и составляет 74.99%. Это связано с переводом категориальных признаков в числовые на работу с которыми и рассчитана линейная регрессия.
##5. Минимальное значение прогноза -  0.005744518567402761
## Максимальное значение прогноза - 0.9937054595749831
